### ISSUE TITLE
staticcheck: (type SA9???) Check if static bit shifts will always clear a value

### ISSUE THREAD

**Issue by klauspost** [dominikh/go-tools#654]
Inspired by [this](https://github.com/klauspost/compress/commit/30596e9f81c65b439a013fab5e467a37ea8cb744#diff-675c122cf030e21c18f8091f10a4a025R197).

Check if a static shift will always clear a value.

```
var v uint8
v >>= 8  // error: v will always be 0.
v >>= 9  // error: v will always be 0.
v <<= 8  // error: v will always be 0.
```

For variable length types (`int`, `uint`) this check *could* be problematic. 

```
// Clear any value above 32 bits if integers are more than 32 bits.
func f(i int) int {
  v := i >> 32
  v = v << 32
  return i-v
}
```
Though this code is dodgy, something like it *could* be used in the wild, so excluding the types with variable length is probably reasonable.

**Comment by dgryski** [comment]
I've seen this bug both in my own code and in other code bases, generally in code that was translated from C. This generally happens due to C's arithmetic promotion rules in which bytes are promoted to ints in the expression.

**Comment by klauspost** [comment]
Yes, I've seen C-code that relies on the (undefined) behaviour that there is an implicit mod register-length. When ported directly to Go this is a problem.

**Comment by dominikh** [comment]
Can do. Though note that `go vet` should already catch this: ` v (8 bits) too small for shift of 8`

**Comment by klauspost** [comment]
Really. I will investigate if this is a problem that go vet is supposed to handle.

**Comment by klauspost** [comment]
Oh, it is because that test isn't included in the subset they run when doing `go test`. :/

**Comment by dominikh** [comment]
I'm reopening the issue; it's a check we can and should implement, after all :-)

**Comment by klauspost** [comment]
Sure :)

**Comment by dominikh** [comment]
This was implemented in e306c661babf7e7dc9f0d784f83d820d7999c20e.
